File: app/src-tauri/src/commands/dev_log.rs
//! Debug-only sink for frontend log forwarding.
//!
//! The dev shell's WebView2 console isn't visible to anything outside the
//! window — F12 only. When the user wants the supervising tool (Claude) to
//! see what the frontend is doing, frontend code calls `invoke("dev_log",
//! ...)` and we append a line to `<repo_root>/.claude-dev.log`. The file is
//! .gitignored and only this command writes to it.
//!
//! The command itself is `#[cfg(debug_assertions)]` and only registered in
//! debug builds (see `lib.rs::generate_handler!`), so release binaries
//! cannot reach it. The frontend stub returns silently when the command
//! isn't registered, so production builds incur no behaviour change.

use std::fs::OpenOptions;
use std::io::Write;
use std::path::PathBuf;
use std::sync::Mutex;

use chrono::Utc;
use serde_json::Value;

use crate::commands::error::CommandError;

/// Resolves the log file path at runtime. `CARGO_MANIFEST_DIR` points at
/// `app/src-tauri/` — going up two levels gives the repo root where Claude
/// is invoked, so the file lands at a predictable location.
fn log_file_path() -> PathBuf {
    let manifest = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    manifest
        .parent()
        .and_then(|p| p.parent())
        .map(|p| p.join(".claude-dev.log"))
        .unwrap_or_else(|| PathBuf::from(".claude-dev.log"))
}

/// Serialised access to the log file. `OpenOptions::append` is itself
/// atomic on most filesystems, but the lock keeps interleaved partial
/// writes from concurrent commands neatly line-separated.
static WRITE_LOCK: Mutex<()> = Mutex::new(());

fn append_line(line: &str) {
    let _guard = WRITE_LOCK.lock();
    let path = log_file_path();
    if let Ok(mut f) = OpenOptions::new().create(true).append(true).open(&path) {
        let _ = writeln!(f, "{line}");
    }
}

/// Called once at app startup so each session is visually demarcated when
/// scrolling through the file. Includes the build identifier so it's clear
/// which build (dev vs prod identity) wrote the surrounding lines.
pub fn log_session_start(identifier: &str) {
    let ts = Utc::now().to_rfc3339();
    append_line(&format!(
        "\n=== session start {ts} :: {identifier} ==="
    ));
}

#[tauri::command]
pub async fn dev_log(
    level: String,
    message: String,
    context: Option<Value>,
) -> Result<(), CommandError> {
    let ts = Utc::now().to_rfc3339();
    let lvl = level.to_uppercase();
    let ctx = match context {
        Some(Value::Null) | None => String::new(),
        Some(v) => format!(" {}", v),
    };
    // tracing pipeline so the line shows up in `yarn dev` stdout too — that
    // way Claude can either tail the yarn output OR read the log file,
    // whichever is more convenient.
    match lvl.as_str() {
        "ERROR" => tracing::error!("[FE] {message}{ctx}"),
        "WARN" => tracing::warn!("[FE] {message}{ctx}"),
        "INFO" => tracing::info!("[FE] {message}{ctx}"),
        _ => tracing::debug!("[FE] {message}{ctx}"),
    }
    append_line(&format!("{ts} [{lvl}] {message}{ctx}"));
    Ok(())
}

File: app/src/lib/devLog.ts
/**
 * Dev-mode log forwarder.
 *
 * Tauri's WebView2 console is only visible via F12 in the running window —
 * not to anything outside the process. This module mirrors `console.log/info/
 * warn/error`, `window.onerror`, and `unhandledrejection` into the Rust
 * `dev_log` command, which appends each line to `<repo_root>/.claude-dev.log`
 * and also re-emits via `tracing` so the lines surface in `yarn dev` stdout.
 *
 * Effect: an external supervisor (e.g. Claude reading the file directly) can
 * see what the frontend is doing without anyone keeping DevTools open.
 *
 * Production safety: the whole module is gated by `import.meta.env.DEV` at
 * the call site (`main.tsx`). Vite tree-shakes the import in production.
 * Even if a release build *did* import it, the Rust `dev_log` command is
 * `#[cfg(debug_assertions)]` and not registered, so invokes would fail and
 * the safe-swallow path kicks in — no behaviour change for users.
 */
import { TauriCommand } from "@recrest/shared";

import { invoke, isTauri } from "@/lib/tauri";
import { setIpcTrace } from "@/lib/tauri/ipcTrace";

type ConsoleLevel = "log" | "info" | "warn" | "error" | "debug";

let installed = false;

/** Best-effort JSON serialisation that survives circular refs, functions,
 *  Error objects, and DOM nodes — enough to capture *intent* for log lines. */
function safeStringify(value: unknown): string {
  if (typeof value === "string") return value;
  if (value instanceof Error) {
    return `${value.name}: ${value.message}\n${value.stack ?? ""}`.trim();
  }
  const seen = new WeakSet<object>();
  try {
    return JSON.stringify(
      value,
      (_key, v: unknown) => {
        if (typeof v === "bigint") return v.toString();
        if (typeof v === "function") return `[Function ${v.name || "anonymous"}]`;
        if (v && typeof v === "object") {
          if (seen.has(v as object)) return "[Circular]";
          seen.add(v as object);
        }
        return v;
      },
      2,
    );
  } catch {
    try {
      return String(value);
    } catch {
      return "[unserialisable]";
    }
  }
}

function joinArgs(args: unknown[]): string {
  return args.map(safeStringify).join(" ");
}

/**
 * Forward a single log line to the Rust sink. The function deliberately
 * does not go through the project's `invoke()` wrapper for two reasons:
 * 1. It must never log itself (would loop the patched console).
 * 2. It must silently swallow `command not found` so the patched console
 *    still works in a half-built dev shell where `dev_log` isn't
 *    registered yet (e.g. the running binary predates the rebuild).
 */
async function forward(level: ConsoleLevel, args: unknown[]): Promise<void> {
  if (!isTauri()) return;
  try {
    await invoke(TauriCommand.DEV_LOG, {
      level,
      message: joinArgs(args),
      context: null,
    });
  } catch {
    // Swallow — see docblock.
  }
}

/** Installs the forwarders. Safe to call multiple times; second call is a
 *  no-op. Returns nothing — the forwarders run for the page lifetime. */
export function installDevLogForwarder(): void {
  if (installed) return;
  installed = true;

  const original: Record<ConsoleLevel, (...args: unknown[]) => void> = {
    log: console.log.bind(console),
    info: console.info.bind(console),
    warn: console.warn.bind(console),
    error: console.error.bind(console),
    debug: console.debug.bind(console),
  };

  const wrap = (level: ConsoleLevel) =>
    function patched(...args: unknown[]): void {
      void forward(level, args);
      original[level](...args);
    };

  console.log = wrap("log");
  console.info = wrap("info");
  console.warn = wrap("warn");
  console.error = wrap("error");
  console.debug = wrap("debug");

  // Surface runtime exceptions that never went through `console.error`
  // (e.g. handler bugs in plain event listeners, sync throws in effects).
  window.addEventListener("error", (e) => {
    void forward("error", [
      `window.onerror: ${e.message}`,
      `at ${e.filename}:${e.lineno}:${e.colno}`,
      e.error instanceof Error ? (e.error.stack ?? "") : "",
    ]);
  });

  // React 18+ swallows render-time errors by default — they flow through
  // error boundaries. Unhandled async rejections still bubble to this
  // listener and are the most common silent-failure mode in IPC paths.
  window.addEventListener("unhandledrejection", (e) => {
    const reason = e.reason as unknown;
    void forward("error", [
      "unhandledrejection",
      reason instanceof Error
        ? `${reason.name}: ${reason.message}\n${reason.stack ?? ""}`
        : safeStringify(reason),
    ]);
  });

  // Force IPC trace on whenever the forwarder runs. The flag also gates
  // a `console.debug("[ipc]", …)` call inside `invoke()` — with the
  // forwarder in place, that call hits our `dev_log` sink, so every
  // `invoke()` (command + args + duration + result/error) lands in
  // `.claude-dev.log`. That's the difference between "the user clicked
  // *something*" and "the user invoked `scan_repos` with paths=[…] and
  // it returned []".
  setIpcTrace(true);

  // Initial marker so the file shows when the SPA finished hydrating.
  void forward("info", [
    `[devLog] forwarder installed at ${new Date().toISOString()} (ipc-trace=on)`,
  ]);
}
